Add encrypted stickers support

surakin opened the pull request:
Fixes https://github.com/ruma/ruma/issues/1819
Fixes https://github.com/mautrix/whatsapp/issues/662
Fixes https://github.com/element-hq/element-x-ios/issues/1047

I implemented a custom deserializer for StickerEventContent to handle the encrypted sticker events generated by several bridges out there.

<!--

PR checklist, not strictly necessary but generally useful unless you're just
fixing a typo or something like that:

- Run `cargo xtask ci` locally before posting the PR
- Documented public API changes in CHANGELOG.md files

-->

surakin committed:
add custom StickerEventContent parser to handle encrypted stickers generated by several bridges

surakin committed:
fix formatting

surakin committed:
use #[derive(Default)] instead

zecakeh commented:
The simpler solution here would be to use `MediaSource`. However we do not deviate from the spec without a `compat-*` feature. A solution would be to have an almost identical `StickerMediaSource`, but the `Encrypted` variant is behind a `compat-encrypted-sticker` feature, with a link to https://github.com/matrix-org/matrix-spec/issues/1667 for the reason of the compat feature.

surakin commented:
I'm not sure I follow. Currently this is not working because the encrypted sticker events don't have the 'url' field so I'm filling it using the url field from the fileinfo struct.
The resulting StickerEventContent struct still follows the spec, even if it was built from a non spec compliant event.

zecakeh commented:
You shouldn't be able to "just" take the URL from the encrypted file. If the file is encrypted, we need the encryption data to decrypt it too, that means all the fields in [`EncryptedFile`](https://docs.ruma.dev/ruma/events/room/struct.EncryptedFile.html).

surakin commented:
Ah I see (my testing environment was a bit of a mess) I'll keep iterating on this :D

surakin committed:
add compat-encrypted-stickers feature

surakin commented:
I added the mediasource and it seems to work well. Not sure if I added the compat flag correctly (my apologies, I'm fairly new to rust :D )
(a few small tweaks are needed in both matrix-sdk and element-x to display this anyway)

surakin committed:
fix formatting

surakin committed:
fix automated tests

surakin committed:
fix formatting and tests

surakin committed:
fix formatting again

surakin committed:
Remove unneeded extra 'url' field in FileInfo

surakin committed:
fix clippy default features test

surakin committed:
Add link to encrypted stickers issue

surakin committed:
Tweak docs and tests

zecakeh commented on crates/ruma-events/Cargo.toml:
```diff
@@ -56,6 +56,9 @@ compat-optional = []
 # Allow TagInfo to contain a stringified floating-point value for the `order` field.
 compat-tag-info = []
 
+# Encrypted stickers
```
Please use the same comment as in the ruma crate.

zecakeh commented on crates/ruma-events/src/sticker.rs:
```diff
@@ -26,11 +31,142 @@ pub struct StickerEventContent {
+    /// The media source
+    #[cfg(not(feature = "compat-encrypted-stickers"))]
+    #[serde(skip)]
+    pub source: MediaSource,
```
This is not what I meant. Since `MediaSource` also contains the `url` field in theory, it should also replace the `url` field.

But since we want the `Encrypted` variant to not be available by default, we should use an almost identical type, with the only difference that the `Encrypted` variant is behind the cargo feature.

zecakeh commented on crates/ruma-events/src/sticker.rs:
```diff
@@ -26,11 +31,142 @@ pub struct StickerEventContent {
+    }
+}
+
+impl<'de> Deserialize<'de> for StickerEventContent {
```
You shouldn't need to implement `Deserialize` for `StickerEventContent`, we should rely on the deserialization of the new `MediaSource`, like we do on `ImageMessageEventContent`.

zecakeh commented on crates/ruma/Cargo.toml:
```diff
@@ -153,6 +154,10 @@ compat-signature-id = ["ruma-signatures?/compat-signature-id"]
 # Allow TagInfo to contain a stringified floating-point value for the `order` field.
 compat-tag-info = ["ruma-events?/compat-tag-info"]
 
+# Encrypted stickers (sent by several bridges, but not quite supported in the spec)
```
```suggestion
# Support encrypted stickers, as sent by several bridges.
```

surakin committed:
Update crates/ruma/Cargo.toml

Co-authored-by: Kévin Commaille <76261501+zecakeh@users.noreply.github.com>

surakin committed:
Update crates/ruma-events/Cargo.toml

surakin committed:
Add StickerMediaSource, drop custom deserializer

surakin committed:
Fix formatting

surakin commented:
Hey you are right, this was much simpler than what I was doing :)

surakin committed:
Fix tests

surakin committed:
Fix test without encryption

zecakeh commented on crates/ruma-events/src/sticker.rs:
```diff
@@ -6,7 +6,33 @@ use ruma_common::OwnedMxcUri;
 use serde::{Deserialize, Serialize};
 
-use crate::room::ImageInfo;
+#[cfg(not(feature = "compat-encrypted-stickers"))]
```
I forgot to tell you in the previous review, the feature should actually be without the `not` everywhere

zecakeh commented on crates/ruma-events/src/sticker.rs:
```diff
@@ -24,13 +50,31 @@ pub struct StickerEventContent {
+    #[cfg(feature = "compat-encrypted-stickers")]
+    pub fn new(body: String, info: ImageInfo, url: OwnedMxcUri) -> Self {
+        Self { body, info, source: StickerMediaSource::Plain(url.clone()) }
+    }
```
Those methods are identical, you don't need to use the feature flag.

zecakeh commented on crates/ruma-events/src/sticker.rs:
```diff
@@ -24,13 +50,31 @@ pub struct StickerEventContent {
+    /// Creates a new `StickerEventContent` with the given body, image info and URL.
+    #[cfg(feature = "compat-encrypted-stickers")]
+    pub fn from_source(body: String, info: ImageInfo, url: OwnedMxcUri) -> Self {
+        Self { body, info, source: StickerMediaSource::Plain(url.clone()) }
```
The first method should be enough, and it doesn't need the feature flag because it would work in both cases. Also I believe that the Rust convention would be to name it `with_source`.

zecakeh commented on crates/ruma-events/src/sticker.rs:
```diff
@@ -6,7 +6,33 @@ use ruma_common::OwnedMxcUri;
+    Encrypted(Box<EncryptedFile>),
+}
+
+impl From<StickerMediaSource> for MediaSource {
```
This is wrong, it will only work if the feature is enabled so it should be behind the feature flag. Also the opposite conversion might make sense.

zecakeh replied:
You didn't handle this comment. Also nitpick, leave an empty line between the impl blocks

surakin replied:
Yeah I didn't quite understand this one :thinking:

surakin replied:
The From implementation can be used with or without the flag

zecakeh replied:
Ah nevermind, it is correct. And a `impl From<MediaSource> for StickerMediaSource` behind the feature would also make sense.

surakin replied:
I didn't have any need for it but I can it for completeness

zecakeh commented on crates/ruma-events/tests/it/sticker.rs:
```diff
@@ -79,7 +79,35 @@ fn content_deserialization() {
+        "info": {},
+    });
+
+    let encrypted_content = from_json_value::<StickerEventContent>(encrypted_json_data).unwrap();
```
This should actually fail if the feature is not enabled. You should probably have something like that:

```rust
#[cfg(not(feature = "compat-encrypted-sticker"))]
{
  from_json_value::<StickerEventContent>(encrypted_json_data).unwrap_err();
}

#[cfg(feature = "compat-encrypted-sticker")]
{
  let encrypted_content = from_json_value::<StickerEventContent>(encrypted_json_data).unwrap();

  assert_eq!(encrypted_content.body, "Upload: my_image.jpg");
  assert_matches!(encrypted_content.source, StickerMediaSource::Encrypted(encrypted_sticker_url));
  assert_eq!(encrypted_sticker_url.url, "mxc://notareal.hs/file");
}

zecakeh commented on crates/ruma-events/src/sticker.rs:
```diff
@@ -6,7 +6,33 @@ use ruma_common::OwnedMxcUri;
+use crate::room::{ImageInfo, MediaSource};
+
+/// The source of a sticker media file.
+#[derive(Clone, Debug, Deserialize, Serialize)]
```
I believe that we should implement the same `Deserialize` as `MediaSource`, to have the same behavior.

surakin committed:
Fix inverted usage of feature flag
Add custom StickerMediaSource deserializer
Clean duplicated code

surakin committed:
Fix formatting

zecakeh commented on crates/ruma-events/src/sticker.rs:
```diff
@@ -4,9 +4,67 @@
+                Err(de::Error::missing_field("url"))
+            }
+            #[cfg(not(feature = "compat-encrypted-stickers"))]
+            StickerMediaSourceJsonRepr { url: None } => Err(de::Error::missing_field("url")),
```
You should be able to simplify this to a single entry:

```suggestion
            StickerMediaSourceJsonRepr {
                url: None,
                #[cfg(feature = "compat-encrypted-stickers")]
                file: None,
            } => {
                Err(de::Error::missing_field("url"))
            }
```

zecakeh commented on crates/ruma-events/src/sticker.rs:
```diff
@@ -4,9 +4,67 @@
+
+/// The source of a sticker media file.
+#[derive(Clone, Debug, Serialize)]
+#[allow(clippy::exhaustive_enums)]
```
This enum needs to be non-exhaustive:

```suggestion
#[non_exhaustive]
```

zecakeh commented on crates/ruma-events/src/sticker.rs:
```diff
@@ -24,13 +82,20 @@ pub struct StickerEventContent {
+    }
+
+    /// Creates a new `StickerEventContent` with the given body, image info, URL, and media source.
+    #[cfg(feature = "compat-encrypted-stickers")]
```
We don't need to put this method behind a feature.

```suggestion
```

zecakeh reviewed (changes requested):
We will also need a breaking change entry in the changelog of the crate.

surakin committed:
Add changelog entry
Simplify StickerMediaSourceJsonRepr
Fix non_exhaustive flag

surakin commented:
Almost there!

surakin committed:
Add conversion from StickerMediaSource to MediaSource under the encrypted-stickers feature

zecakeh commented on crates/ruma-events/CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+Breaking changes:
+
+ - Add support for encrypted stickers as sent by several bridges under the flag `compat-encrypted-stickers`
```
This could be listed as an improvement. The breaking change here is that the `url` field of `StickerEventContent` has been replaced with the `source` field that is a `StickerMediaSource`.

surakin committed:
Update changelog

zecakeh commented:
Thanks for this!
